ODIN_II/SRC/partial_map.cpp

```diff
@@ -899,132 +896,113 @@ void instantiate_GE(nnode_t* node, operation_list type, short mark, netlist_t* n
-        /* record the size of the shift */
-        shift_size = node->related_ast_node->children[1]->types.vnumber->get_value();
-    } else {
-        shift_size = 0;
-        error_message(NETLIST, node->loc, "%s\n", "Odin only supports constant shifts at present");
-    }
+void instantiate_shift(nnode_t* node, operation_list type, short mark, netlist_t* netlist) {
+    /*  
+     *   Create mux 2:1
+     *   data1 = SHIFT first_input
+     *   data2 = SHIFT first_input shifted right or left by pow(2,i)
+     *   selector = SHIFT second_input[i]
+     */
 
-    buf_node = make_1port_gate(BUF_NODE, width, width, node, mark);
+    oassert(node->related_ast_node->children[1]->type != NUMBERS);
 
-    if (type == SL || type == ASL) {
-        /* IF shift left */
-
-        /* connect inputs to outputs */
-        for (i = 0; i < width - shift_size; i++) {
-            // connect higher output pin to lower input pin
-            remap_pin_to_new_node(node->input_pins[i], buf_node, i + shift_size);
-        }
-
-        /* connect ZERO to outputs that don't have inputs connected */
-        for (i = 0; i < shift_size; i++) {
-            // connect 0 to lower outputs
-            add_input_pin_to_node(buf_node, get_zero_pin(netlist), i);
-        }
+    nnode_t** muxes;
+    npin_t** input_pins;
+    npin_t** output_pins;
+    int in_out_port_width;
+
+    input_pins = node->input_pins;
+    output_pins = input_pins;
+    in_out_port_width = node->output_port_sizes[0];
+    muxes = (nnode_t**)vtr::malloc(sizeof(nnode_t*) * (node->input_port_sizes[1]));
+
+    for (int i = 0; i < in_out_port_width; i++) {
+        int power_two_by_i = shift_left_value_with_overflow_check(0x1, i, node->related_ast_node->loc);
+
+        input_pins = output_pins;
+
+        muxes[i] = make_3port_gate(MULTI_PORT_MUX, 1, in_out_port_width, in_out_port_width, in_out_port_width, node, mark);
+        // connect related pin of second_input to related multiplexer as a selector
+        remap_pin_to_new_node(node->input_pins[in_out_port_width + i], muxes[i], 0);
+
+        // connect the output of the previous stage mux as the first input to the next stage mux
+        for (int j = 0; j < in_out_port_width; j++)
+            remap_pin_to_new_node(input_pins[j], muxes[i], j + (1));
+
+        if (type == SR || type == ASR) {
+            /* 
+             * Logical variable shift right 
+             * or 
+             * arithmatic variable shift right
+             */
+
+            // shift by pow(2,i) and connect the output of the previous stage mux as the second input to the next stage mux
+            for (int j = power_two_by_i; j < in_out_port_width; j++)
+                remap_pin_to_new_node(input_pins[j],
+                                      muxes[i],
+                                      j - power_two_by_i + (1 + in_out_port_width));
+
+            // connect the sign bit of the previous output as extension bits
+            int pad_bit = in_out_port_width - 1;
```

Comment:
pad bit only is used if the input is signed, you must check for that

Reply:
> > So, how can we find is the input port signed or not using the netlist node
> 
> The pins should already have been allocated space and plugged on the generic "ASR" or "SR" node before, all you have to do is build the barrel shifter. (The actual implement not its IO)

Pins here should be already fine in terms of sign, appropriate extension bits are added to input

Reply:
i see that now! thanks for making the changes there